### PR TITLE
Refactor decode_integer

### PR DESCRIPTION
**Pull request by Ian-Foote**
I'd like to benchmark these four commits to see what speed-ups, if any, they provide.

### CODE REVIEW

**Comment by Lukasa** [comment]
So I'm in 100% agreement: I think we need to construct a benchmark to get a feel for how this affects the speed of processing. I'm inclined to say that this is almost certainly an improvement, but by how much.

**Comment by Ian-Foote** [comment]
I don't know what a suitable benchmark would look like. Do you have any ideas or suggestions how to find out?

**Comment by Lukasa** [comment]
Frankly, I think the best benchmarks are going to be `timeit` based for a large number of loops. The function is sufficiently simple that if you check what numbers of prefix bits are used throughout the codebase and then run multiple instances of `timeit` using integers of various different sizes for each of those prefix bits, it should help us get a feel for what's going on. For what it's worth, the codebase uses 4, 5, 6, and 7 as values for `prefix_bits`.

**Comment by Ian-Foote** [comment]
What does typical `data` look like? I tried using the values in `test_hpack.py`, but I didn't see a big difference. If the typical data is longer I'd expect an improvement.

**Comment by Lukasa** [comment]
The typical data can vary in length, but generally is only a few bytes long. The biggest possible integers in regular use in HPACK are likely to be of the order of, say, 30,000. When encoded with a four bit prefix, that will only produce a four-byte-long output from `encode_integer`. It should be noted that the encoded integers have three uses in HPACK: the first is as an index into a table that rarely gets more than a few hundred values long; the second is as a length of a header field name or value, which could be on the order of tens or hundreds of kB but is surely not larger; and finally as a storage for how large the header table is allowed to be, which again is unlikely to be larger than a few MB. Either way, anything more than about 6 bytes is stunningly unlikely, and anything more than 8 is downright unreasonable (an 8 byte integer can store integer values up to 18,446,744,073,709,551,615, and the odds of us getting a header field value longer than *that* are pretty unlikely!).

**Comment by Lukasa** [comment]
Ok, I've written a micro-benchmark based on decoding `2**20` encoded with a different number of prefix bits from 4 to 7, on different interpreters. The code for this is available [in this gist](https://gist.github.com/Lukasa/31564f9d4bed0faf94c10b2f21831706), which I recommend people run themselves to reproduce the results. I've arranged them into four tables, below. Underneath these tables I have provided some analysis.

##### 4 Prefix Bits

| Platform | Old Run 1 | Old Run 2 | Old Run 3 | New Run 1 | New Run 2 | New Run 3 | Best Old Run | Best New Run | Improvement | Improvement (%) |
|----------|-----------|-----------|-----------|-----------|-----------|-----------|--------------|--------------|-------------|-----------------|
| CPython 2.7.12 | 1.6426 | 1.6183 | 1.6214 | 1.5508 | 1.5182 | 1.5469 | 1.6183 | 1.5182 | 0.1002 | 6.19% |
| CPython 3.3.6 | 2.1991 | 2.1819 | 2.1731 | 1.6059 | 1.6128 | 1.6073 | 2.1731 | 1.6059 | 0.5673 | 26.10% |
| CPython 3.4.5 | 2.2584 | 2.3067 | 2.2535 | 1.6561 | 1.6354 | 1.6700 | 2.2535 | 1.6354 | 0.6181 | 27.43% |
| CPython 3.5.2 | 2.3044 | 2.2992 | 2.3478 | 1.8173 | 1.7586 | 1.7645 | 2.2992 | 1.7586 | 0.5406 | 23.51% |
| PyPy 2.7.10 | 0.0985 | 0.1129 | 0.0888 | 0.0867 | 0.0766 | 0.0772 | 0.0888 | 0.0766 | 0.0123 | 13.81% |

##### 5 Prefix Bits

| Platform | Old Run 1 | Old Run 2 | Old Run 3 | New Run 1 | New Run 2 | New Run 3 | Best Old Run | Best New Run | Improvement | Improvement (%) |
|----------|-----------|-----------|-----------|-----------|-----------|-----------|--------------|--------------|-------------|-----------------|
| CPython 2.7.12 | 1.6417 | 1.6292 | 1.6187 | 1.5459 | 1.5320 | 1.5342 | 1.6187 | 1.5320 | 0.0867 | 5.36% |
| CPython 3.3.6 | 2.1754 | 2.2493 | 2.1857 | 1.6287 | 1.6161 | 1.6317 | 2.1754 | 1.6161 | 0.5593 | 25.71% |
| CPython 3.4.5 | 2.2433 | 2.2600 | 2.2219 | 1.6772 | 1.6682 | 1.6552 | 2.2219 | 1.6552 | 0.5667 | 25.51% |
| CPython 3.5.2 | 2.3656 | 2.3542 | 2.3927 | 1.7979 | 1.7701 | 1.7976 | 2.3542 | 1.7701 | 0.5841 | 24.81% |
| PyPy 2.7.10 | 0.1007 | 0.0871 | 0.0884 | 0.0816 | 0.0742 | 0.0674 | 0.0871 | 0.0674 | 0.0197 | 22.58% |

##### 6 Prefix Bits

| Platform | Old Run 1 | Old Run 2 | Old Run 3 | New Run 1 | New Run 2 | New Run 3 | Best Old Run | Best New Run | Improvement | Improvement (%) |
|----------|-----------|-----------|-----------|-----------|-----------|-----------|--------------|--------------|-------------|-----------------|
| CPython 2.7.12 | 1.6342 | 1.6229 | 1.6657 | 1.5422 | 1.5462 | 1.5322 | 1.6229 | 1.5322 | 0.0908 | 5.59% |
| CPython 3.3.6 | 2.2087 | 2.1845 | 2.1919 | 1.6296 | 1.6904 | 1.6267 | 2.1845 | 1.6267 | 0.5577 | 25.53% |
| CPython 3.4.5 | 2.2161 | 2.2661 | 2.6481 | 1.9037 | 1.8310 | 1.6977 | 2.2161 | 1.6977 | 0.5184 | 23.39% |
| CPython 3.5.2 | 2.3305 | 2.3529 | 2.3247 | 1.8093 | 1.7884 | 1.7918 | 2.3247 | 1.7884 | 0.5363 | 23.07% |
| PyPy 2.7.10 | 0.0978 | 0.0803 | 0.0821 | 0.0925 | 0.0782 | 0.0777 | 0.0803 | 0.0777 | 0.0026 | 3.22% |

##### 7 Prefix Bits

| Platform | Old Run 1 | Old Run 2 | Old Run 3 | New Run 1 | New Run 2 | New Run 3 | Best Old Run | Best New Run | Improvement | Improvement (%) |
|----------|-----------|-----------|-----------|-----------|-----------|-----------|--------------|--------------|-------------|-----------------|
| CPython 2.7.12 | 1.6669 | 1.6095 | 1.6800 | 1.5634 | 1.5466 | 1.5255 | 1.6095 | 1.5255 | 0.0840 | 5.22% |
| CPython 3.3.6 | 2.1846 | 2.1756 | 2.2164 | 1.6323 | 1.6578 | 1.6177 | 2.1756 | 1.6177 | 0.5579 | 25.64% |
| CPython 3.4.5 | 2.2648 | 2.2614 | 2.2511 | 1.6744 | 1.8941 | 1.7115 | 2.2511 | 1.6744 | 0.5767 | 25.62% |
| CPython 3.5.2 | 2.3498 | 2.3508 | 2.3914 | 1.8056 | 1.7867 | 1.7854 | 2.3498 | 1.7854 | 0.5644 | 24.02% |
| PyPy 2.7.10 | 0.1124 | 0.0865 | 0.1014 | 0.1066 | 0.0688 | 0.0834 | 0.0865 | 0.0688 | 0.0176 | 20.40% |


So, what have we learned here?

1. PyPy is stupidly good at this kind of workload, and while it sees some marginal improvement from this change the improvement is highly variable and entirely down to good luck on the run. It can be as low as a few percentage points, or as high as 20% faster with the new workload. Regardless, this provides another TL;DR for anyone who hadn't realised: **use PyPy**!
2. Python 2.7 seems to be dominated by other work, so this algorithm change doesn't help it that much. I haven't profiled it, but I *suspect* that the dominant workload is actually the conversion of single-byte-string to integer in `to_byte`. There is an improvement here, but it's pretty minor: a few %age points.
3. Python 3 is *way* slow at this workload compared to Python 2.7 in the original case! Wow! This change provides a huge benefit to Python 3, somehow removing a really slow operation and replacing it with one Python 3 is much faster at doing. This changes brings Python 3 to be just slightly slower than Python 2, which is a good change.
4. Interestingly, longer bit lengths don't see incrementally better improvement, at least not in a way that outweighs noise. It seems like the improvement is more or less a flat 25% on Python 3 and about a flat 5% on Python 2 (including PyPy).

TL;DR: this newer variant is faster in all cases and *much faster* on Python 3, good work @Ian-Foote!

**Comment by Lukasa** [comment]
Before we merge this, I'd like to check in some tests that use hypothesis to confirm that our encoder/decoder logic round-trips. That will help us better validate that these functions are really being refactored, and aren't about to encounter some subtle behavioural changes.

**Comment by Ian-Foote** [comment]
I can reproduce the speed-ups with your benchmark.

**Comment by Ian-Foote** [comment]
I've rebased as suggested in #64.

**Comment by Lukasa** [comment]
@Ian-Foote Per our [one of the team](http://python-hyper.org/en/latest/one-of-the-team.html) policy, you'll find an invite to the Python Hyper organisation on GitHub in your inbox. Welcome to the team @Ian-Foote!
